src/Std/Data/DTreeMap/Raw.lean

```diff
@@ -0,0 +1,278 @@
+
+Lemmas about the operations on `Std.DTreeMap.Raw` will be available in the module
+`Std.Data.DTreeMap.RawLemmas`.
+-/
+
+set_option autoImplicit false
+set_option linter.missingDocs true
+
+universe u v
+
+variable {α : Type u} {β : α → Type v} {cmp : α → α → Ordering}
+private local instance : Coe (Type v) (α → Type v) where coe γ := fun _ => γ
+
+namespace Std
+
+namespace DTreeMap
+
+/--
+Dependent tree maps without a bundled well-formedness invariant, suitable for use in nested
+inductive types. The well-formedness invariant is called `Raw.WF`. When in doubt, prefer `DTreeMap`
+over `DTreeMap.Raw`. Lemmas about the operations on `Std.DTreeMap.Raw` are available in the
+module `Std.Data.DTreeMap.RawLemmas`.
+
+A tree map stores an assignment of keys to values. It depends on a comparator function that
+defines an ordering on the keys and provides efficient order-dependent queries, such as retrieval
+of the minimum or maximum.
+
+To ensure that the operations behave as expected, the comparator function `cmp` should satisfy
+certain laws that ensure a consistent ordering:
+
+* If `a` is less than (or equal) to `b`, then `b` is greater than (or equal) to `a`
+and vice versa (see the `OrientedCmp` typeclass).
+* If `a` is less than or equal to `b` and `b` is, in turn, less than or equal to `c`, then `a`
+is less than or equal to `c` (see the `TransCmp` typeclass).
+
+Keys for which `cmp a b = Ordering.eq` are considered the same, i.e., there can be only one entry
+with key either `a` or `b` in a tree map. Looking up either `a` or `b` always yields the same entry,
+if any is present. The `get` operations of the _dependent_ tree map additionally require a
+`LawfulEqCmp` instance to ensure that `cmp a b = .eq` always implies `a = b`, so that their
+respective value types are equal.
+
+To avoid expensive copies, users should make sure that the tree map is used linearly.
+
+Internally, the tree maps are represented as size-bounded trees, a type of self-balancing binary
+search tree with efficient order statistic lookups.
+-/
+structure Raw (α : Type u) (β : α → Type v) (_cmp : α → α → Ordering := by exact compare) where
+  /-- Internal implementation detail of the tree map. -/
+  inner : Internal.Impl α β
+
+namespace Raw
+
+/--
+Well-formedness predicate for tree maps. Users of `DTreeMap` will not need to interact with
+this. Users of `DTreeMap.Raw` will need to provide proofs of `WF` to lemmas and should use lemmas
+like `WF.empty` and `WF.insert` (which are always named exactly like the operations they are about)
+to show that map operations preserve well-formedness. The constructors of this type are internal
+implementation details and should not be accessed by users.
+-/
+structure WF (t : Raw α β cmp) where
+  /-- Internal implementation detail of the tree map. -/
+  out : letI : Ord α := ⟨cmp⟩; t.inner.WF
+
+instance {t : Raw α β cmp} : Coe t.WF (letI : Ord α := ⟨cmp⟩; t.inner.WF) where
+  coe h := h.out
```

Comment:
I'm not sure I understand where exactly in this file this is used.

Reply:
It will be used by the proofs for the lemmas, simplifying the proof terms a bit. I tend to keep it there although it is currently unused, but I can remove it until it is used if you wish.

Reply:
You can keep it if you like.